packages/effect/src/DateTime.ts

```diff
@@ -0,0 +1,1991 @@
+)
+
+/**
+ * Get the current time as a `DateTime.Zoned`, using the `CurrentTimeZone`.
+ *
+ * @since 3.6.0
+ * @category current time zone
+ * @example
+ * import { DateTime, Effect } from "effect"
+ *
+ * Effect.gen(function* () {
+ *   // will use the "Europe/London" time zone
+ *   const now = yield* DateTime.nowInCurrentZone
+ * }).pipe(DateTime.withCurrentZoneNamed("Europe/London"))
+ */
+export const nowInCurrentZone: Effect.Effect<Zoned, never, CurrentTimeZone> = Effect.flatMap(
+  now,
+  setZoneCurrent
+)
+
+/**
+ * Create a Layer from the given time zone.
+ *
+ * @since 3.6.0
+ * @category current time zone
+ */
+export const layerCurrentZone = (zone: TimeZone): Layer.Layer<CurrentTimeZone> => Layer.succeed(CurrentTimeZone, zone)
+
+/**
+ * Create a Layer from the given time zone offset.
+ *
+ * @since 3.6.0
+ * @category current time zone
+ */
+export const layerCurrentZoneOffset = (offset: number): Layer.Layer<CurrentTimeZone> =>
+  Layer.succeed(CurrentTimeZone, zoneMakeOffset(offset))
+
+/**
+ * Create a Layer from the given IANA time zone identifier.
+ *
+ * @since 3.6.0
+ * @category current time zone
+ */
+export const layerCurrentZoneNamed = (zoneId: string): Layer.Layer<CurrentTimeZone, IllegalArgumentException> =>
+  Layer.effect(CurrentTimeZone, zoneMakeNamedEffect(zoneId))
+
+/**
+ * Create a Layer from the systems local time zone.
+ *
+ * @since 3.6.0
+ * @category current time zone
+ */
+export const layerCurrentZoneLocal: Layer.Layer<CurrentTimeZone> = Layer.sync(
+  CurrentTimeZone,
+  zoneMakeLocal
+)
+
+// =============================================================================
+// mapping
+// =============================================================================
+
+const makeZonedFromAdjusted = (adjustedMillis: number, zone: TimeZone): Zoned => {
+  const offset = zone._tag === "Offset" ? zone.offset : calculateNamedOffset(adjustedMillis, zone)
+  return makeZonedProto(adjustedMillis - offset, zone)
+}
+
+const offsetRegex = /([+-])(\d{2}):(\d{2})$/
+const parseOffset = (offset: string): number | null => {
+  const match = offsetRegex.exec(offset)
+  if (match === null) {
+    return null
+  }
+  const [, sign, hours, minutes] = match
+  return (sign === "+" ? 1 : -1) * (Number(hours) * 60 + Number(minutes)) * 60 * 1000
+}
+
+const calculateNamedOffset = (adjustedMillis: number, zone: TimeZone.Named): number => {
+  const offset = zone.format.formatToParts(adjustedMillis).find((_) => _.type === "timeZoneName")?.value ?? ""
+  if (offset === "GMT") {
+    return 0
+  }
+  const result = parseOffset(offset)
+  if (result === null) {
+    // fallback to using the adjusted date
+    return zonedOffset(makeZonedProto(adjustedMillis, zone))
+  }
+  return result
+}
+
+/**
+ * Modify a `DateTime` by applying a function to the underlying `Date`.
+ *
+ * The `Date` will first have the time zone applied if possible, and then be
+ * converted back to a `DateTime` within the same time zone.
+ *
```

Comment:
Does `mutate` and `mutateUtc` need a comment here explaining the meaning/intention of the function name? i.e., that it is safe and expected to use `Date`'s mutative methods because it's been pre-cloned before being passed to the function